OSDOCS#6402: GCP on with multi-arch compute machines support (#63991)
openshift/openshift-docs: modules/multi-architecture-modify-machine-set-gcp.adoc

            value: <custom_metadata_value>
          kind: GCPMachineProviderSpec
          machineType: n1-standard-4 <5>
          metadata:
            creationTimestamp: null
          networkInterfaces:
          - network: <infrastructure_id>-network
            subnetwork: <infrastructure_id>-worker-subnet
          projectID: <project_name> <6>
          region: us-central1 <7>
          serviceAccounts:
          - email: <infrastructure_id>-w@<project_name>.iam.gserviceaccount.com
            scopes:
            - https://www.googleapis.com/auth/cloud-platform
          tags:
            - <infrastructure_id>-worker
          userDataSecret:
            name: worker-user-data
          zone: us-central1-a
----
<1> Specify the infrastructure ID that is based on the cluster ID that you set when you provisioned the cluster. If you have the OpenShift CLI installed, you can obtain the infrastructure ID by running the following command:
+
[source,terminal]
----
$ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
----
<2> Specify the role node label to add.
<3> Specify the path to the image that is used in current compute machine sets. 

> // the OCP docs reference the GCP Marketplace image, not sure if that is supported on ARM64

Reviewer: i don't believe the market place images are supported, but I guess the same query that we use to get AMis for AWS can be used here to fetch the appropriate gcp images from the rhcos.json file in the installer

Author: Is there a specific command that can do that similar to AWS for GCP? Or is it something completely different?

Reviewer: it should be similar to what we have for aws today: https://docs.openshift.com/container-platform/4.13/post_installation_configuration/multi-architecture-configuration.html#multi-architecture-modify-machine-set-aws_multi-architecture-configuration.

The command to get the arm64 image for gcp will have to be slightly tweaked, so it should be:
```
oc get configmap/coreos-bootimages /
	  -n openshift-machine-config-operator /
	  -o jsonpath='{.data.stream}' | jq /
	  -r '.architectures.<arch>.images.gcp.name'
```
Although I'm not entirely sure the deployment directly uses this image as a bootimage or creates a temporary image. @aleskandro can you or your QE colleagues verify this?

Reviewer 2: Yes, the command above is fine.

The path to the image takes this form:
```php
projects/${project}/global/images/${image_name}
```

with image_name as above and project as in  

```
oc get configmap/coreos-bootimages / -n openshift-machine-config-operator / -o jsonpath='{.data.stream}' | jq / -r '.architectures.<arch>.images.gcp.project'
```

In this case arch is the GNU arch (aarch64, x86_64, others...)

Reviewer 2: nit in the command: `/` vs `\`

          machineType: n1-standard-4 <5>
          metadata:
            creationTimestamp: null
          networkInterfaces:
          - network: <infrastructure_id>-network
            subnetwork: <infrastructure_id>-worker-subnet
          projectID: <project_name> <6>
          region: us-central1 <7>
          serviceAccounts:
          - email: <infrastructure_id>-w@<project_name>.iam.gserviceaccount.com
            scopes:
            - https://www.googleapis.com/auth/cloud-platform
          tags:
            - <infrastructure_id>-worker
          userDataSecret:
            name: worker-user-data
          zone: us-central1-a
----
<1> Specify the infrastructure ID that is based on the cluster ID that you set when you provisioned the cluster. If you have the OpenShift CLI installed, you can obtain the infrastructure ID by running the following command:
+
[source,terminal]
----
$ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
----
<2> Specify the role node label to add.
<3> Specify the path to the image that is used in current compute machine sets. You can run the following command to access the and ARM64 supported image. 

> + 

Author: @Prashanth684 @aleskandro Hello! Thank you for the info! Does this look right?

Reviewer: looking at @aleskandro's comment [here](https://github.com/openshift/openshift-docs/pull/63991#discussion_r1331352050 ), it looks like we have to provide the information about the path which includes the project and the name. You have the command to get the project. You need to add the command to get the name and also mention the form of the path

Author: Ah I see now, I misunderstood the original comment, thank you!

Author: Okay, I think I updated it right!

Reviewer: similar to the project you need a snippet for the <image_name>:

```
To access the image name, run the following command:

oc get configmap/coreos-bootimages /
  -n openshift-machine-config-operator /
  -o jsonpath='{.data.stream}' | jq /
  -r '.architectures.aarch64.images.gcp.name'
```

          machineType: n1-standard-4 <5>
          metadata:
            creationTimestamp: null
          networkInterfaces:
          - network: <infrastructure_id>-network
            subnetwork: <infrastructure_id>-worker-subnet
          projectID: <project_name> <6>
          region: us-central1 <7>
          serviceAccounts:
          - email: <infrastructure_id>-w@<project_name>.iam.gserviceaccount.com
            scopes:
            - https://www.googleapis.com/auth/cloud-platform
          tags:
            - <infrastructure_id>-worker
          userDataSecret:
            name: worker-user-data
          zone: us-central1-a
----
<1> Specify the infrastructure ID that is based on the cluster ID that you set when you provisioned the cluster. If you have the OpenShift CLI installed, you can obtain the infrastructure ID by running the following command:
+
[source,terminal]
----
$ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
----
<2> Specify the role node label to add.
<3> Specify the path to the image that is used in current compute machine sets. You can use the following commands to access a ARM64 supported image. 
+

> To access the image name, follow the example path in your terminal: 

Reviewer 2: ```suggestion
The image path should follow the following format:
```

+
[source,terminal]
----

> projects/${project}/global/images/<image_name>

Reviewer 2: ```suggestion
projects/<project_name>/global/images/<image_name>
```

Reviewer: i don't see this change..are we doing this?

----

> To access the image name, run the following command: 

Author: My apologies for my confusion on this one, its just these two items that are needed right? @aleskandro

Reviewer 2: Let me show you... The data we are retrieving with the command below comes from [this json](https://github.com/openshift/installer/blob/master/data/data/coreos/rhcos.json#L235-L239).
```json
"architectures": {
   "aarch64": {
      "images": {
        "gcp": {
          "release": "415.92.202309142014-0",
          "project": "rhcos-cloud",
          "name": "rhcos-415-92-202309142014-0-gcp-aarch64"
        }
  ......

```

users must set the path to the image as in row 103. From the example above, it would be :

`projects/rhcos-cloud/global/images/rhcos-415-92-202309142014-0-gcp-aarch64`

I think the options we have are two:

#### Document two similar commands to retrieve (i) the project name, (ii) the image name:

project name:
```bash
oc get configmap/coreos-bootimages \
  -n openshift-machine-config-operator \
  -o jsonpath='{.data.stream}' | jq \
  -r '.architectures.aarch64.images.gcp.project'

[example output] 
rhcos-cloud
```

image name:

```bash
oc get configmap/coreos-bootimages \
  -n openshift-machine-config-operator \
  -o jsonpath='{.data.stream}' | jq \
  -r '.architectures.aarch64.images.gcp.name'

[example output]  
rhcos-415-92-202309142014-0-gcp-aarch64
```

#### Document only one command

we can also let the users print the json and specify that they need to use the `project` and `name` fields to fill the path ` projects/<project>/global/images/<name>` based on the json output of the command:

```bash
oc get configmap/coreos-bootimages \
  -n openshift-machine-config-operator \
  -o jsonpath='{.data.stream}' | jq \
  -r '.architectures.aarch64.images.gcp'

[example output]  
        "gcp": {
          "release": "415.92.202309142014-0",
          "project": "rhcos-cloud",
          "name": "rhcos-415-92-202309142014-0-gcp-aarch64"
        }
```

The first way gives users two redundant commands that provide specific output for each missing part of the path URI. The second way gives one only command and requires the users to look at the json output to fille the image path.

What do you think is the best to provide?

Author: That actually helps a bunch, thank you Alessandro. I like having just the one output and I can do some rewording so that users know what part they need.

Author: @aleskandro Hello! I just updated the PR and I'm more confident about the updates, WDYT?

Reviewer: i think the example output above should also be included mentioning that the `project` and `name` field should be used respectively (as Alessandro already mentioned):

```
[example output]  
        "gcp": {
          "release": "415.92.202309142014-0",
          "project": "rhcos-cloud",
          "name": "rhcos-415-92-202309142014-0-gcp-aarch64"
        }
```

Reviewer 2: +1 to @Prashanth684

Author: Just added the example output and did some rewording! Thank you @Prashanth684 and @aleskandro, should be all set now

+ 
[source,terminal]

> ----
> oc get configmap/coreos-bootimages /
>   -n openshift-machine-config-operator /
>   -o jsonpath='{.data.stream}' | jq /
>   -r '.architectures.aarch64.images.gcp.project'
> ----

Reviewer 2: ```suggestion
----
oc get configmap/coreos-bootimages \
  -n openshift-machine-config-operator \
  -o jsonpath='{.data.stream}' | jq \
  -r '.architectures.aarch64.images.gcp.project'
----
```

//Module included in the following assembly
//
//post_installation_configuration/configuring-multi-arch-compute-machines/creating-multi-arch-compute-nodes-gcp.adoc

> :_content-type: PROCEDURE
> [id="multi-architecture-modify-machine-set-gcp_{context}"]
> 
> = Adding an ARM64 compute machine set to your GCP cluster 

Reviewer 3: ```suggestion
:_content-type: PROCEDURE
[id="multi-architecture-modify-machine-set-gcp_{context}"]
= Adding an ARM64 compute machine set to your GCP cluster 
```
Suggestion - remove unnecessary space

> To configure a cluster with multi-architecture compute machines, you must create a GCP ARM64 compute machine set. This adds ARM64 compute nodes to your cluster so that your cluster has multi-architecture compute machines.

Reviewer 3: ```suggestion
To configure a cluster with multi-architecture compute machines, you must create a GCP ARM64 compute machine set. This adds ARM64 compute nodes to your cluster.
```
Suggestion - you could probably leave this duplication of the first sentence off.

Author: Great suggestion, yeah its a bit redundant when included

.Prerequisites 

> * You installed the OpenShift CLI (`oc`). 

Reviewer 3: ```suggestion
* You installed the {oc-first}. 
```
There's an attribute for this now 😉

Author: Ooo thank ya!

* You used the installation program to create an AMD64 single-architecture AWS cluster with the multi-architecture installer binary.

.Procedure

> * Create and modify a compute machine set, this will control the ARM64 compute nodes in your cluster. 

Reviewer 3: ```suggestion
* Create and modify a compute machine set that controls the ARM64 compute nodes in your cluster:
```
Edit suggestion for present tense and formatting using `:`

        value:
          apiVersion: gcpprovider.openshift.io/v1beta1
          canIPForward: false
          credentialsSecret:
            name: gcp-cloud-credentials
          deletionProtection: false
          disks:
          - autoDelete: true
            boot: true
            image: <path_to_image> <3>
            labels: null
            sizeGb: 128
            type: pd-ssd
          gcpMetadata: <4>
          - key: <custom_metadata_key>
            value: <custom_metadata_value>
          kind: GCPMachineProviderSpec
          machineType: n1-standard-4 <5>
          metadata:
            creationTimestamp: null
          networkInterfaces:
          - network: <infrastructure_id>-network
            subnetwork: <infrastructure_id>-worker-subnet
          projectID: <project_name> <6>
          region: us-central1 <7>
          serviceAccounts:
          - email: <infrastructure_id>-w@<project_name>.iam.gserviceaccount.com
            scopes:
            - https://www.googleapis.com/auth/cloud-platform
          tags:
            - <infrastructure_id>-worker
          userDataSecret:
            name: worker-user-data
          zone: us-central1-a
----

> <1> Specify the infrastructure ID that is based on the cluster ID that you set when you provisioned the cluster. If you have the OpenShift CLI installed, you can obtain the infrastructure ID by running the following command:

Reviewer 3: ```suggestion
<1> Specify the infrastructure ID that is based on the cluster ID that you set when you provisioned the cluster. You can obtain the infrastructure ID by running the following command:
```
The `oc` command is specified as a prerequisite, so we should be able to assume they have it for this step?

+
[source,terminal]
----
$ oc get -o jsonpath='{.status.infrastructureName}{"\n"}' infrastructure cluster
----
<2> Specify the role node label to add.
<3> Specify the path to the image that is used in current compute machine sets. You need the project and image name for your path to image. 
+
To access the project and image name, run the following command: 
+ 
[source,terminal]
----
$ oc get configmap/coreos-bootimages \ 
  -n openshift-machine-config-operator \ 
  -o jsonpath='{.data.stream}' | jq \ 
  -r '.architectures.aarch64.images.gcp'
----
+
.Example output 
[source,terminal]
----
  "gcp": {
    "release": "415.92.202309142014-0",
    "project": "rhcos-cloud",
    "name": "rhcos-415-92-202309142014-0-gcp-aarch64"
  }
----
Use the `project` and `name` parameters from the output to create the path to image field in your machine set. The path to the image should follow the following format:
+
[source,terminal]
----

> $ projects/<project>/global/images/<image_name>

Author: I added that change here @Prashanth684

----

> <4> Optional: Specify custom metadata in the form of a `key:value`` pair. For example use cases, see the GCP documentation for link:https://cloud.google.com/compute/docs/metadata/setting-custom-metadata[setting custom metadata].

Reviewer 3: ```suggestion
<4> Optional: Specify custom metadata in the form of a `key:value` pair. For example use cases, see the GCP documentation for link:https://cloud.google.com/compute/docs/metadata/setting-custom-metadata[setting custom metadata].
```

> <5> Specify an ARM64 supported machine type. For more information, refer to "Tested instance types for GCP 64-bit ARM"

Reviewer 3: ```suggestion
<5> Specify an ARM64 supported machine type. For more information, refer to "Additional resources".
```
I'd maybe just point to the Additional resources section. Reading through this at first I thought, _where would I find those docs_? 🙂
